device/samples/edge_downstream_device.js

```diff
@@ -0,0 +1,91 @@
+// 2) The edge device hostname is the hostname set in the config.yaml of the Edge device
+//    to which this sample will connect to.
+//
+// The resulting string should look like the following
+//  "HostName=<iothub_host_name>;DeviceId=<device_id>;SharedAccessKey=<device_key>;GatewayHostName=<edge device hostname>"
+var connectionString = '[Downstream device IoT Edge connection string]';
+
+// Path to the Edge "owner" root CA certificate
+var edge_ca_cert_path = '[Path to Edge CA certificate]';
+
+// fromConnectionString must specify a transport constructor, coming from any transport package.
+var client = Client.fromConnectionString(connectionString, Protocol);
+
+var connectCallback = function (err) {
+  if (err) {
+    console.error('Could not connect: ' + err.message);
+  } else {
+    console.log('Client connected');
+    client.on('message', function (msg) {
+      console.log('Id: ' + msg.messageId + ' Body: ' + msg.data);
+      // When using MQTT the following line is a no-op.
+      client.complete(msg, printResultFor('completed'));
+      // The AMQP and HTTP transports also have the notion of completing, rejecting or abandoning the message.
+      // When completing a message, the service that sent the C2D message is notified that the message has been processed.
+      // When rejecting a message, the service that sent the C2D message is notified that the message won't be processed by the device. the method to use is client.reject(msg, callback).
+      // When abandoning the message, IoT Hub will immediately try to resend it. The method to use is client.abandon(msg, callback).
+      // MQTT is simpler: it accepts the message by default, and doesn't support rejecting or abandoning a message.
+    });
+
+    // Create a message and send it to the IoT Hub every second
+    var sendInterval = setInterval(function () {
+      var windSpeed = 10 + (Math.random() * 4); // range: [10, 14]
+      var temperature = 20 + (Math.random() * 10); // range: [20, 30]
+      var humidity = 60 + (Math.random() * 20); // range: [60, 80]
+      var data = JSON.stringify({ deviceId: 'myFirstDownstreamDevice', windSpeed: windSpeed, temperature: temperature, humidity: humidity });
+      var message = new Message(data);
+      message.properties.add('temperatureAlert', (temperature > 28) ? 'true' : 'false');
+      console.log('Sending message: ' + message.getData());
+      client.sendEvent(message, printResultFor('send'));
+    }, 2000);
+
+    client.on('error', function (err) {
+      console.error(err.message);
+    });
+
+    client.on('disconnect', function () {
+      clearInterval(sendInterval);
+      client.removeAllListeners();
+      client.open(connectCallback);
+    });
+  }
+};
+
+// Provide the Azure IoT device client via setOptions with the X509
+// Edge root CA certificate that was used to setup the Edge runtime
+var options = {
+  ca : fs.readFileSync(edge_ca_cert_path, 'utf-8').toString(),
```

Comment:
>.toString() [](start = 50, length = 11)

The string SHOULD be unnecessary.  I believe I don't use it in my code.

Reply:
I looked at this for reference: https://github.com/Azure/azure-iot-sdk-node/blob/c1ae87bf7c769a7193d3add07213a8b57c579540/device/samples/simple_sample_device_x509.js#L66

I can try without and re-test.